File: solutions/bit-manipulation/alternatives/231-power.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Main solution for Problem 231: Power of Two
 *
 * @param {number} n - An integer
 * @return {boolean} - True if n is a power of two, false otherwise
 *
 * Time Complexity: O(1)
 * Space Complexity: O(1)
 */
function solve(n) {
    // Power of 2 has exactly one bit set
    // n & (n-1) removes the rightmost set bit
    // If n is power of 2, this results in 0
    // Also check n > 0 to handle negative numbers and zero
    return n > 0 && (n & (n - 1)) === 0;
}

/**
 * Test cases for Problem 231: Power
 */
function testSolution() {
    console.log('Testing 231. Power');

    // Test case 1: Power of 2
    const result1 = solve(1);
    const expected1 = true;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Power of 2
    const result2 = solve(16);
    const expected2 = true;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Not power of 2
    const result3 = solve(3);
    const expected3 = false;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Zero
    const result4 = solve(0);
    const expected4 = false;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: Negative number
    const result5 = solve(-16);
    const expected5 = false;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    // Test case 6: Large power of 2
    const result6 = solve(1024);
    const expected6 = true;
    console.assert(result6 === expected6, `Test 6 failed: expected ${expected6}, got ${result6}`);

    console.log('All test cases passed for 231. Power!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 231. Power ===');
    console.log('Category: Bit Manipulation');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on bit manipulation concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: docs/solutions/bit-manipulation/231-power.py
"""
# Difficulty: Medium

# 231. Power

Given a problem that demonstrates key concepts in Bit Manipulation.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[input description]</dd>
<dt>Output:</dt>
<dd>[output description]</dd>
<dt>Explanation:</dt>
<dd>[explanation]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of bit manipulation concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply bit manipulation methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages bit manipulation principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

class Solution:
    def isPowerOfTwo(self, n: int) -> bool:
        """
        Check if a number is a power of two using bit manipulation.

        Args:
            n: Integer to check

        Returns:
            True if n is a power of 2, False otherwise

        Time Complexity: O(1) - constant time bit operations
        Space Complexity: O(1) - no extra space needed
        """
        # Power of 2 has exactly one bit set
        # n & (n-1) removes the rightmost set bit
        # For power of 2, this should result in 0
        return n > 0 and (n & (n - 1)) == 0

    def solve(self, n: int) -> bool:
        """Wrapper method for consistency with template."""
        return self.isPowerOfTwo(n)

def test_solution():
    """
    Test cases for 231. Power of Two.
    """
    solution = Solution()

    # Test case 1: Power of 2 - 1
    result = solution.solve(1)
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Power of 2 - 16
    result = solution.solve(16)
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: Not power of 2 - 3
    result = solution.solve(3)
    expected = False
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 4: Power of 2 - 4
    result = solution.solve(4)
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 5: Not power of 2 - 5
    result = solution.solve(5)
    expected = False
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 6: Power of 2 - 1024
    result = solution.solve(1024)
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 7: Negative number
    result = solution.solve(-16)
    expected = False
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 8: Zero
    result = solution.solve(0)
    expected = False
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 9: Power of 2 - 2
    result = solution.solve(2)
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    test_values = [1, 2, 3, 4, 16, 17, 1024]
    print(f"Solution for 231. Power of Two:")
    for val in test_values:
        result = solution.solve(val)
        print(f"{val}: {result}")
